### DISCUSSION TITLE
Separating Classes & Using Inheritance With UserControl

### DISCUSSION THREAD

**Discussion by AaronCatolico** [flet-dev/flet#1055] [Q&A]
I have two separate classes ('Container1' & 'Container2'). I'd like the Container2 class to be able to inherit Container1 controls (variables and functions), but every class in the end will be called from the 'main.py' file to add all classes controls to the page. Here's a very simple example below of Container1 & Container2 classes. I'd like to be able to click the button in Container2 class container to change the color of Container1 container:

```
import flet as ft

class Container1(ft.UserControl):

    def change_color1(self, e):
        self.c1.bgcolor='orange'
        self.c1.update()

    def build(self):
        self.c1 = ft.Container(width=200, height=200, bgcolor='red', alignment=ft.alignment.center, margin=10)
        self.btn1 = ft.TextButton(text='COUNT', style=ft.ButtonStyle(bgcolor='green', color='white'), on_click=self.change_color1)
        self.c1.content = ft.Column([self.btn1], alignment=ft.MainAxisAlignment.CENTER)
        return ft.Column([self.c1])


class Container2(Container1):
    
    def change_color2(self, e):
        pass
        #when clicking btn2 in Container2, the background color of c1 in Container1 class will turn blue

    def build(self):
        self.c2 = ft.Container(width=200, height=200, bgcolor='red', alignment=ft.alignment.center, margin=10)
        self.btn2 = ft.TextButton(text='COUNT', style=ft.ButtonStyle(bgcolor='green', color='white'), on_click=self.change_color2)
        self.c2.content = ft.Column([self.btn2], alignment=ft.MainAxisAlignment.CENTER)
        return ft.Column([self.c2])



def main(page):
    page.window_width = 600
    page.window_height = 580
    page.window_center()
    page.bgcolor = '#333333'
    page.padding = 0
    page.window_always_on_top = True
    page.update()

    main_container = ft.Row([Container1(), Container2()])

    page.add(main_container)

ft.app(target=main)
```

**Comment by iron3oxide** [comment]
Understood, but what's your question?

**Reply by AaronCatolico** [reply]
Oh, I thought it was obvious or already basically included in the post. What I'm asking is how do I inherit the Container1 class into the Container2 class to change the c1 bgcolor to blue by clicking the btn2 in c2 in Container2 class.

**Reply by iron3oxide** [reply]
If you inherit as described in the code block, you should be able to access and therefore modify c1 in `change_color2()` simply by referencing `self.c1`. I'm not sure it's the best way to design this, but it should work.

**Reply by AaronCatolico** [reply]
Here's what I've tried below, but I'm still not able to access c1 to change it's background color:

```
    def change_color2(self, e):
        self.c1.bgcolor='blue'
        self.c1.update()
```

I keep getting the error message:
`AttributeError: 'Container2' object has no attribute 'c1'. Did you mean: 'c2'?`

**Reply by iron3oxide** [reply]
Move the `self.c1 = ft.Container(width=200, height=200, bgcolor='red', alignment=ft.alignment.center, margin=10)` line into an `__init__` method. The way it is set up now, `c1` doesn't exist until `build()` is called.

**Reply by AaronCatolico** [reply]
Can you show me an example of what you mean? How do you move the self.c1 line into a __init__ method?

**Reply by iron3oxide** [reply]
```python
import flet as ft

class Container1(ft.UserControl):
    def __init__(self):
        super().__init__()
        self.c1 = ft.Container(width=200, height=200, bgcolor='red', alignment=ft.alignment.center, margin=10)
        self.btn1 = ft.TextButton(text='COUNT', style=ft.ButtonStyle(bgcolor='green', color='white'), on_click=self.change_color1)
        self.c1.content = ft.Column([self.btn1], alignment=ft.MainAxisAlignment.CENTER)

    def change_color1(self, e):
        self.c1.bgcolor='orange'
        self.c1.update()

    def build(self):
        return ft.Column([self.c1])


class Container2(Container1):
    def __init__(self):
        super().__init__()
        self.c2 = ft.Container(width=200, height=200, bgcolor='red', alignment=ft.alignment.center, margin=10)
        self.btn2 = ft.TextButton(text='COUNT', style=ft.ButtonStyle(bgcolor='green', color='white'), on_click=self.change_color2)
        self.c2.content = ft.Column([self.btn2], alignment=ft.MainAxisAlignment.CENTER)
    
    def change_color2(self, e):
        pass
        #when clicking btn2 in Container2, the background color of c1 in Container1 class will turn blue

    def build(self):
        return ft.Column([self.c2])



def main(page):
    page.window_width = 600
    page.window_height = 580
    page.window_center()
    page.bgcolor = '#333333'
    page.padding = 0
    page.window_always_on_top = True
    page.update()

    main_container = ft.Row([Container1(), Container2()])

    page.add(main_container)

ft.app(target=main)
```

The `__init__` method is sort of the constructor of your class. It takes arguments used to construct instances of your class and is generally the place where instance variables (like `self.c1`) should be initialized. That's why I went ahead and changed the rest of the code as well.

**Reply by AaronCatolico** [reply]
Here's what I've tried, but I'm still getting the error `AssertionError: Control must be added to the page first.`

```
import flet as ft


class Container1(ft.UserControl):
    def __init__(self):
        super().__init__()
        self.c1 = ft.Container(width=200, height=200, bgcolor='red', alignment=ft.alignment.center, margin=10)
        self.btn1 = ft.TextButton(text='COUNT', style=ft.ButtonStyle(bgcolor='green', color='white'), on_click=self.change_color1)
        self.c1.content = ft.Column([self.btn1], alignment=ft.MainAxisAlignment.CENTER)

    def change_color1(self, e):
        self.c1.bgcolor = 'orange'
        self.c1.update()

    def build(self):
        return ft.Column([self.c1])


class Container2(Container1):
    def __init__(self):
        super().__init__()
        self.c2 = ft.Container(width=200, height=200, bgcolor='blue', alignment=ft.alignment.center, margin=10)
        self.btn2 = ft.TextButton(text='COUNT', style=ft.ButtonStyle(bgcolor='green', color='white'), on_click=self.change_color2)
        self.c2.content = ft.Column([self.btn2], alignment=ft.MainAxisAlignment.CENTER)

    def change_color2(self, e):
        self.c1.bgcolor='blue'
        self.c1.update()
        # when clicking btn2 in Container2, the background color of c1 in Container1 class will turn blue

    def build(self):
        return ft.Column([self.c2])


def main(page):
    page.window_width = 600
    page.window_height = 580
    page.window_center()
    page.bgcolor = '#333333'
    page.padding = 0
    page.window_always_on_top = True
    page.update()

    main_container = ft.Row([Container1(), Container2()])

    page.add(main_container)


ft.app(target=main)
```

**Reply by iron3oxide** [reply]
Could you paste in the full error message? It would be helpful to find out what caused the error. Regardless of that though, I wonder why you define `build()` methods but never actually use them.

**Reply by AaronCatolico** [reply]
UserControl must implement build() method that is called to build control's UI and should return a single Control instance or a List of controls. 
https://flet.dev/docs/guides/python/user-controls

This is what is shown in total error messages:

```
C:\Users\Aaron\AppData\Local\Programs\Python\Python310\python.exe C:/Users/Aaron/Documents/Python/Flet-Testing/main.py
Exception in thread Thread-7 (change_color2):
Traceback (most recent call last):
  File "C:\Users\Aaron\AppData\Local\Programs\Python\Python310\lib\threading.py", line 1016, in _bootstrap_inner
    self.run()
  File "C:\Users\Aaron\AppData\Local\Programs\Python\Python310\lib\threading.py", line 953, in run
    self._target(*self._args, **self._kwargs)
  File "C:\Users\Aaron\Documents\Python\Flet-Testing\main.py", line 28, in change_color2
    self.c1.update()
  File "C:\Users\Aaron\AppData\Local\Programs\Python\Python310\lib\site-packages\flet_core\control.py", line 260, in update
    assert self.__page, "Control must be added to the page first."
AssertionError: Control must be added to the page first.
```

**Reply by iron3oxide** [reply]
Just read the `UserControl` docs and it seems you have to call `self.update()` instead of `self.c1.update()`.

**Reply by AaronCatolico** [reply]
> Just read the `UserControl` docs and it seems you have to call `self.update()` instead of `self.c1.update()`.

I've already read the docs, that's how I ended up with the starting code that I originally posted, but the docs doesn't show any examples of the class inheritance which is why I ended up here. Also, I've already tried to use 'self.update()', but that still doesn't change the situation with the errors.

**Reply by iron3oxide** [reply]
I see. Do you need to use Container1 and/or Container2 somewhere else in your code or are they really just intended for main_container?

**Reply by AaronCatolico** [reply]
I really would like my classes separated to keep different sections of controls that all do their own thing, but will sometimes need to interact. E.g. Container1 will have it's own list of controls, Container2 will have it's own list of controls, but sometimes Container2 may need to make changes to Container1 or vice versa.

**Reply by iron3oxide** [reply]
Understandable, but will you (re)use them in different parts of your UI or just in the main_container?

**Reply by AaronCatolico** [reply]
I think that I'd really only use them in the 'main_container' to call 1 instance of each class, then add the 'main_container' to the page.

**Reply by iron3oxide** [reply]
In that case, I'd advise you to keep the two Container classes completely separate (no inheritance) and define a MainContainer class which contains an instance of both container classes. You could then either pass each container constructor the instance of the other container class or do the interaction logic in the MainContainer by employing an Observer pattern (MainContainer registers a callback with each container, which is called by them when e.g. the button is clicked).

**Reply by AaronCatolico** [reply]
Yes, I've thought of a Main class of some kind as the one that will receive all of the controls from separate classes, but wasn't sure how to implement it all. In this case, how would that be written?

**Reply by iron3oxide** [reply]
My advice on how to write this depends on many factors, the most important one being if you intend to use the program in a production setting and maintain it for a longer period of time. The complexity of the program (will it have multiple views and thus need routing, route protection and cross-view state) is another important factor to consider.

**Reply by AaronCatolico** [reply]
No, it won't have multiple views and need any routing. If you know how to create a basic example of a '`main.py`' file that can pull in other classes from other files/modules such as a '`control1.py`' file with class '**Control1**' (it's properties & functions), and a '`control2.py`' file with '**Control2**' class, along with it's properties and functions, then have them work together, that would most likely be the best structure for this project.
